### DISCUSSION TITLE
Allow long polling as basis of subscriptions

### DISCUSSION THREAD

**Discussion by revero-doug** [trpc/trpc#4263] [Ideas]
I know, I know, websockets (and SSE) are much more hip and modern, but hear me out...

Both kafka and redis use long polling, and for good reason. Long-lived connections are fragile and don't scale well. Long polling has several desirable properties when compared with websockets:

1. when the system is under too much load, a shorter max connection time yields a more graceful degredation of service. clients are more likely to experience intermittent failures while they roll the dice on opening a connection, whereas with long-lived connections, some clients are lucky and potentially doing nothing with the connection, while others are unable to establish a connection. See the ["Universal Law of Computational Scalability" (USL)](https://en.wikipedia.org/wiki/Neil_J._Gunther#Universal_Law_of_Computational_Scalability); from this follows that breaking the session into several short-lived connections, the mean time to successfully acquire a connection goes down. More importantly, becomes more predictable and evenly distributed among clients, so even if it's wasteful, it's easily remedied by horizontal scaling and has a better failure mode when it's insufficiently scaled.
2. persistent sessions (whether via long polling or a long-lived ws connection) often aren't very saturated, i.e. there's a lot of waiting with a connection open, and that consumes a lot of memory
3. since it's regular old http sans upgrade, it's relatively impervious to corporate firewalls, default loadbalancer policies, naggle's algorithm -- all the thing that make websockets "hard"

**Comment by revero-doug** [comment]
you can enable configuring a number of parameters to mitigate the cost of opening a connection; here are a few basic ones:

- `minPollIntervalMS`: throttles polling frequency
- `timeoutMS`: deadline at which any retrieved messages sent in a response
- `minBatchSize`: number of messages to collect before sending a response
    - if timeoutMS runs out before the min batch size is met, the batch is sent in the response

**Comment by Carnewal** [comment]
I definitely think we should have the option.

Similar to this: Server-Sent Events could be used as an alternative subscription mechnism.

https://github.com/trpc/trpc/issues/544

**Comment by jedzej** [comment]
I like this idea, would be helpful also with next.js where creating websocket requires you to eject from the defaults and create custom server.

**Comment by noseworthy** [comment]
I think it'd be awesome if the `httpSubscriptionLink` could fall-back to long polling. We're having a hell of a time getting SSE working in some corporate networks. It's a nightmare. A graceful fallback to polling like what `socket.io` does would be awesome.
